src/plugins/matrixops/hsplit_operation.cpp

```diff
@@ -0,0 +1,267 @@
+            extract_numeric_value_dimensions(args[0], name_, codename_);
+
+        std::size_t num_rows = matrix_dims[0];
+        std::size_t num_cols = matrix_dims[1];
+
+        std::size_t block_dims =
+            extract_numeric_value_dimension(args[1], name_, codename_);
+
+        std::vector<std::pair<std::size_t, std::size_t>> ranges;
+
+        if (block_dims < 1)    // Scalar tiling
+        {
+            phylanx::ir::node_data<double> val =
+                extract_numeric_value(args[1], name_, codename_);
+
+            std::size_t num_blocks = val.scalar();
+            ranges.reserve(num_blocks);
+
+            if (num_blocks <= 0)
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "hsplit_operation::eval",
+                    generate_error_message(
+                        "the hsplit_operation primitive can not split "
+                        "matrices/vectors into fewer blocks than one"));
+            }
+            if (num_blocks > num_cols)
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "hsplit_operation::eval",
+                    generate_error_message(
+                        "the hsplit_operation primitive can not split "
+                        "matrices/vectors into more blocks than there "
+                        "are columns"));
+            }
+            if (num_cols % num_blocks != 0)
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "hsplit_operation::eval",
+                    generate_error_message(
+                        "the hsplit_operation primitive can not split "
+                        "matrices/vectors unevenly"));
+            }
+            std::size_t block_size = num_cols / num_blocks;
+
+            std::size_t index(num_rows);
+            std::size_t tmp_block_num(num_blocks);
+
+            for (int i = 0; i < num_cols; i += block_size)
+            {
+                std::pair<std::size_t, std::size_t> range(i, i + block_size);
+                ranges.push_back(range);
+            }
+        }
+        else    // Index-based tiling
+        {
+            auto&& indices = extract_node_data<T>(std::move(args[1]));
+            auto v = indices.vector();
+
+            if (v.size() > 0)
+            {
+                for (int i = 0; i < v.size(); i++)
+                {
+                    if (static_cast<std::size_t>(v[i]) > num_cols)
+                        v[i] = num_cols;
+                }
+
+                ranges.reserve(v.size() + 1);
+
+                // Include from 0 to first index
+                ranges.push_back(std::make_pair<std::size_t, std::size_t>(
+                    std::size_t(0), std::size_t(v[0])));
+
+                // Make a list of pairs from the passed index list
+                for (int i = 0; i < v.size() - 1; i++)
+                {
+                    ranges.push_back(std::make_pair<std::size_t, std::size_t>(
```

Comment:
FWIW, you should be able to simply write `std::make_pair(…)`, the types will be deduced from the arguments.